src/main/java/org/spongepowered/api/data/value/Value.java

```diff
@@ -0,0 +1,235 @@
+ * {@link ValueContainer}. The actual value may come from various sources of
+ * the {@link ValueContainer}, but usually it's a generic dynamic system for
+ * being able to fetch values from object fields without having to know the
+ * type of {@link Class} of the {@link ValueContainer}, the getters and
+ * setters for that particular value. The driving force behind this is that
+ * instead of having a traditional hierarchical structure of data that is
+ * possible to be retrieved from an {@link Entity}, {@link Living}, etc.,
+ * all that is required is <pre>{@code container.supports(Keys.HEALTH) ?
+ * container.get(Keys.HEALTH).get() : 0 }</pre> where the container is simply
+ * a {@link ValueContainer}, nothing more, nothing less.
+ *
+ * <p>The advantage of this is that now, these various known and unknown
+ * {@link Mutable}s can be retrieved by simple java generics:
+ * {@link ValueContainer#getValue(Key)}. While having a {@link Mutable} for
+ * something so primitive as the current health of a {@link Living} entity,
+ * the power is wielded when a {@link Mutable} can be offered up to multiple
+ * {@link ValueContainer}s without worrying about whether it's supported or not,
+ * or getting the right cast information.</p>
+ *
+ * @param <E> The type of element wrapped by this value
+ */
+public interface Value<E> {
+
+    /**
+     * Gets the held value. Usually all held values are "filled" and not
+     * "defaulted", in some cases however, the default and existing value
+     * are one and the same. Occasionally there are multiple value types
+     * that are not "singular" in the sense, like {@link CollectionValue}s,
+     * or {@link MapValue}s that contain complex types. In these collection
+     * types, usually the default will be an empty collection.
+     *
+     * @return The held value
+     */
+    E get();
+
+    /**
+     * Gets the default value. There is always a default value, however,
+     * usability of the default value may be questionable in certain
+     * circumstances.
+     *
+     * @return The default value
+     */
+    E getDefault();
+
+    /**
+     * Gets the direct value. Since some values may be absent for various
+     * reasons, the {@link #get()} would return the {@link #getDefault()} when
+     * necessary.
+     *
+     * @return The direct value, if available
+     */
+    Optional<E> getDirect();
+
+    /**
+     * Gets the key for this {@link Value}.
+     *
+     * @return The key for this value
+     */
+    Key<? extends Value<E>> getKey();
+
+    /**
+     * Retrieves a mutable form of this value. Due to the vague nature of the
+     * value itself, some cases can already provide a {@link Mutable} instance
+     * where this would simply return itself. In other cases, where the retrieved
+     * value is an {@link Immutable} instance, a new mutable value is created
+     * with the same key, values, and defaults.
+     *
+     * @return A mutable value
+     */
+    Mutable<E> asMutable();
```

Comment:
Please add ad method like `asMutableCopy()` that explcitly creates a new mutable copy.
(This method could be default implemented in the mutable variant to call `copy()` and in the immutable variant by calling to mutable.)

The goal of that method is to explicitly create a copy that can only be modified in the following code and not from the callee without unnecessary object creation.
Also it is more readable then `source.asImmutble().asMutable()` or `source.asMutable().copy()`.

Reply:
You already get copies of objects when retrieved from either `DataHolder`s or `ValueContainer`s (`DataManipulator` extends `ValueContainer`), so calling `asMutable()` is already going to either be a copy from the original container, or a new copy from an immutable value. Adding `asMutableCopy()` only really adds unnecessary steps for what's already intended as a safeguard for developers to use otherwise.

Reply:
Ah ok, on the surface I was thinking Value was going to act as a lens / accessor to immutable / mutable values, But I guess that's more Key's purpose. That's why I was worried about asMutable being exposed on the immutable Value, as if I pass an immutable value to a method, I don't expect it to be able to modify the underlying object.